Test/patterns/PatternMatching.dfy

```diff
@@ -131,6 +131,14 @@ method NewTypeTest() returns (r: int) {
   return;
 }
 
+datatype PairOfNumbers = P(int, int)
+method M(p: PairOfNumbers) returns (r:int) {
+  match p
+  case P(11, 10) => r := 0;
+  case P(11, a) => r := 1;
+  case P(10, 11) => r := 2;
+  case _ => r := 3;
+}
```

Comment:
Good. This tests that no "redundant case" error is generated.

To additionally test that the verifier behaves appropriately on this example, add some postconditions (phrased in some way that doesn't use `match`). For example, add destructors to `PairOfNumbers`, like:

    datatype PairOfNumbers = P(0: int, 1: int)

and then add

    ensures p.0 == 11 && p.1 == 10 ==> r == 0
    ensures p.0 == 11 && p.1 != 10 ==> r == 1
    ensures p.0 == 11 || p.1 == 11 || r == 3

Another way to test the verifier would be to add an `assert a != 10;` on L138 or to add

    assert p.1 == 11 ==> p.0 != 10 && p.1 != 11;

to L140.

Then, to test the compiler, add a couple of calls to `M` from `Main`.

Once you have the interesting test method (here, `M`), it's easy to add these additional tests, and it's good to do so habitually.